### PR TITLE
Feature enumerable - full code with full tests

### PR DESCRIPTION
**Pull request by canriquez**
This project adds my own version of enumerable methods onto the existing Enumerable module. In this project, I considered the full method definition included in https://devdocs.io/ruby~2.6/enumerable, to replicate the same behavior in all defined possible conditions.

Implemented methods are as follow:

#my_each
#my_each_with_index
#my_select
#my_all?
#my_any?
#my_none?
#my_count
#my_map
#my_inject
#my_inject: Using #my_inject to build #multiply_els method
#my_map method, modification to take procs
#my_map refactoring so it takes procs and blocks
Each method is tested within the same code, after module definition.

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress @canriquez 
You have made tremedous progress in this project given it's your first PR. We need to fix a few failing edge cases before we can merge and proceed. Check my comments for suggestions.

### Martin Afani

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+        elsif !arr[i].nil? && arr[i] != false
+          any_h[:t_e] += 1
+        end
+      end
+      any_h[:t_rxp], any_h[:t_p] = param_reg(param, arr[i], any_h[:t_rxp], any_h[:t_p], false)
+    end
+    any_element_true(arr.length, any_h[:t_b_e], any_h[:t_e], any_h[:t_rxp], any_h[:t_p])
+
+    # if we reach this step and all elements are thruty, we exit true.
+  end
```
Comment (whiz25) [inline main.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+    0.upto(arr.length - 1) do |i|
+      if block_given?
+        any_h[:true_block_elements] += 1 if yield arr[i]
+      end
+      any_h[:true_block_elements] = no_block_count(any_h[:true_block_elements], block_given?, arr[i])
+      any_h[:t_rxp], any_h[:true_pattern] = param_reg(param, arr[i], any_h[:t_rxp], any_h[:true_pattern], false)
+    end
+    any_h[:true_block_elements] >= 1 || any_h[:t_rxp] >= 1 || any_h[:true_pattern] >= 1
+  end
```
Comment (whiz25) [inline main.rb]: - [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array.my_any? == true_array.any? #true
```
- [x]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words.my_any?(Integer) == words.any?(Integer) #true
```
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+        if block_given?
+          any_h[:f_b_e] += 1 unless yield arr[i]
+        elsif arr[i].nil? || arr[i] == false
+          any_h[:f_e] += 1
+        end
+      end
+      any_h[:f_rxp], any_h[:f_p] = param_reg(param, arr[i], any_h[:f_rxp], any_h[:f_p], any_h[:who])
+    end
+    any_element_true(arr.length, any_h[:f_b_e], any_h[:f_e], any_h[:f_rxp], any_h[:f_p])
+  end
```
Comment (whiz25) [inline main.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        total[0] = value - value
+      elsif conditions(i.zero?, block_given?, !value.is_a?(Numeric))
+        total[0] = value.clear
+      end
+      total[i + 1] = yield total[i], value
+      i += 1
+    end
+    total[-1]
+  end
+end
```
Comment (whiz25) [inline main.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```

File: main.rb (hunk 1 above)
Comment (canriquez) [reply main.rb]: Got it. Thanks for the hint. This is fixed now.

File: main.rb (hunk 2 above)
Comment (canriquez) [reply main.rb]: A)  This evaluation is working for me: true_array.my_any? == true_array.any? #true
irb(main):477:0> true_array
=> [true, true, true]
irb(main):478:0> true_array.my_any?
=> true
irb(main):479:0> true_array.any?
=> true
irb(main):480:0> true_array.any? == true_array.my_any?
=> true

B) **This evaluation now works. Now fixed:** words.my_any?(Integer) == words.any?(Integer) #true
irb(main):700:0> words
=> ["alo", "mama", "hey"]
irb(main):701:0> words.my_any?(Integer) 
=> false
irb(main):702:0> words.any?(Integer)
=> false
irb(main):703:0> words.my_any?(Integer) == words.any?(Integer)
=> true
c) This evaluation works on my end: words.my_any?(/z/) == words.any?(/z/)
 words
=> ["alo", "mama", "hey"]
irb(main):705:0> words.my_any?(/z/)
=> false
irb(main):706:0> words.any?(/z/)
=> false
irb(main):707:0> words.my_any?(/z/) == words.any?(/z/)
=> true

D) this evaluation now works on my end: words.my_any?('cat') == words.any?('cat')

irb(main):708:0> words.my_any?('cat')
=> false
irb(main):709:0> words.any?('cat')
=> false
irb(main):710:0> words.my_any?('cat') == words.any?('cat')
=> true
irb(main):711:0> words
=> ["alo", "mama", "hey"]

File: main.rb (hunk 3 above)
Comment (canriquez) [reply main.rb]: This method is fixed on my end now for the suggested cases:
irb(main):198:0> array = Array.new(100) { rand(0...9) }
=> [0, 0, 8, 6, 7, 7, 0, 8, 3, 8, 6, 8, 3, 0, 4, 5, 0, 3, 6, 5, 8, 2, 3, 4, 4, 7, 3, 4, 5, 6, 0, 7, 5, 8, 7, 6, 3, 8, 4, 6, 8, 5, 6, 3, 4, 7, 6, 6, 4, 0, 2, 3, 7, 7, 2, 1, 1, 8, 6, 3, 4, 8, 7, 3, 5, 3, 5, 4, 2, 1, 4, 0, 7, 2, 6, 4, 0, 3, 8, 4, 6, 3, 3, 2, 8, 2, 8, 2, 5, 0, 4, 7, 2, 3, 4, 6, 4, 4, 2, 7]
irb(main):199:0> array.my_none?(String) == true
=> true
irb(main):200:0> words = %w[dog door rod blade]
=> ["dog", "door", "rod", "blade"]
irb(main):201:0> words.my_none?(5) == words.none?(5)
=> true
irb(main):202:0> words.my_none?(5) 
=> true
irb(main):203:0> words.none?(5)
=> true
irb(main):204:0>

File: main.rb (hunk 4 above)
Comment (canriquez) [reply main.rb]: This method is now working on my end
irb(main):990:0> array = Array.new(100) { rand(0...9) }
=> [7, 1, 7, 3, 1, 0, 8, 6, 2, 6, 1, 4, 1, 3, 8, 7, 1, 6, 3, 5, 4, 0, 6, 7, 8, 0, 7, 7, 7, 8, 6, 6, 5, 7, 4, 6, 4, 0, 0, 8, 3, 3, 1, 0, 0, 5, 4, 8, 8, 5, 7, 0, 4, 2, 0, 1, 5, 4, 1, 7, 4, 1, 1, 8, 1, 5, 6, 8, 4, 4, 4, 6, 2, 2, 8, 6, 5, 4, 1, 5, 4, 7, 2, 3, 7, 2, 1, 6, 6, 4, 5, 4, 7, 3, 5, 4, 3, 0, 5, 6]
irb(main):991:0> array.my_inject(:+)
=> 417
irb(main):992:0> array.inject(:+)
=> 417
irb(main):993:0> array.my_inject(:+) == array.inject(:+) 
=> true
irb(main):994:0>

**Comment by canriquez** [comment]
All requested changes implemented. I am resubmitting this pull-request.
thanks for your help Afani Martin.

**Review by fatymahmed — CHANGES_REQUESTED** [review]
## Hi @canriquez 
Great job making the suggested changes👏👏, however, there are a few that you might have missed.
- Check my comments below for a detailed review of your methods.

### Status: Changes Required 🛑

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+
+    0.upto(arr.length - 1) do |i|
+      if block_given?
+        any_h[:true_block_elements] += 1 if yield arr[i]
+      end
+      any_h[:t_rxp], any_h[:true_pattern] = param_reg(param, arr[i], any_h[:t_rxp], any_h[:true_pattern], false)
+      any_h[:true_block_elements] = no_block_count(any_h[:true_block_elements], block_given?, arr[i], param)
+    end
+    any_h[:true_block_elements] >= 1 || any_h[:t_rxp] >= 1 || any_h[:true_pattern] >= 1
+  end
```
Comment (fatymahmed) [inline main.rb:102]: For `my_any?`, handle the case when no block or argument is given returns true if at least one of the collection is not false or nil

- Built-in method
`[nil, false, nil, false].any?` returns `false`

- Your method
`[nil, false, nil, false].my_any?` returns `true`

- Check this [document](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-any-3F) for detailed information on `any?` method.

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+        total, param = inject_total_param(total, param)
+        total[i + 1] = total[i].send(param[0], value)
+      else
+        total[i + 1] = yield total[i], value
+      end
+
+      i += 1
+    end
+    total[-1]
+  end
```
Comment (fatymahmed) [inline main.rb:211]: For `my_inject` handle the case when a block is passed with a binary operation, it should combine all elements of enum by applying the binary operation.

- For built-in method
 `%w[dog door rod blade].inject{ |memo, word| memo.length > word.length ? memo : word }` returns `blade`

- Your method gives the following error

> FrozenError:
       can't modify frozen String

Comment (fatymahmed) [reply main.rb:211]: - Check this [doc](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-inject) for more information on `inject` method.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+p [10, 20, 30, 40, 50].inject(5, :+)
+print '[10, 20, 30, 40, 50].my_inject(5,:+) #==>: '
+p [10, 20, 30, 40, 50].my_inject(5, :+)
+puts ''
+
+print '[1, 2, 3, 4, 5].inject(5,:*) #==>: '
+p [1, 2, 3, 4, 5].inject(5, :*)
+print '[1, 2, 3, 4, 5].my_inject(5,:*) #==>: '
+p [1, 2, 3, 4, 5].my_inject(5, :*)
+puts ''
```
Comment (fatymahmed) [inline main.rb]: - Please remove this code, you could place it on a separate file for testing purposes

File: main.rb (hunk 6 above)
Comment (canriquez) [reply main.rb:211]: Hi, I cannot replicate this behavior. 
If I run the code in the command line (IRB) I get the following results. 

irb(main):222:0> %w[dog door rod blade].my_inject{ |memo, word| memo.length > word.length ? memo : word } 
=> "blade"
irb(main):223:0> %w[dog door rod blade].inject{ |memo, word| memo.length > word.length ? memo : word } 
=> "blade"

And so, I don't see an issue here. Can you be more specific on now my_inject is failing?

File: main.rb (hunk 5 above)
Comment (canriquez) [reply main.rb:102]: Thanks for this feedback. this method is now fixed on my end.
irb(main):448:0> [nil, false, nil, false].any?
=> false
irb(main):449:0> [nil, false, nil, false].my_any?
=> false
irb(main):450:0>

File: main.rb (hunk 7 above)
Comment (canriquez) [reply main.rb]: I now have moved the tests to a new file.

File: main.rb (hunk 6 above)
Comment (fatymahmed) [reply main.rb:211]: Hi, So sorry about this.
You are right, the error isn't caused by your method but by the `frozen_string_literal: true` comment.

**Comment by Anna-Myzukina** [comment]
### Good job! :+1: 

## Status [Approved]

* Good job on rectifying the issues raised in the previous reviews :1st_place_medal: . Your methods are passing are behaving! :100: 
#### Go ahead and merge to master.
#### Goodluck in your next project Happy coding <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Review by Anna-Myzukina — APPROVED** [review]

**Comment by canriquez** [comment]
Thank you @whiz25, @fatymahmed  and Anna-Myzukina .

**Pull request merged**